### PR TITLE
feat(minifier): support `property_write_side_effects` option to drop unused property assignments

### LINKED ISSUE #14207: minifier: drop constructed functions/objects
Found this while bundling Ripple. This is required for parity with Rollup/Terser/SWC.

```js
function A() {}
A.from = () => {}
const B = {}
B.from = () => {}
```
should be able to minify into nothing.

New assumptions would be required to accomplish this, per @sapphi-red:
> These require `Object.prototype.from` / `Function.prototype.from` to not be a setter with a sideeffect

### PR DESCRIPTION
**Pull request by Dunqing**
## Summary

Add `property_write_side_effects` option to `TreeShakeOptions` (default `true`). When set to `false`, property assignments like `A.from = () => {}` are considered side-effect-free, enabling the minifier to drop declarations and fresh value bindings with property assignments when unused.

### Examples

```js
// Function declaration + property assignments → eliminated
function A() {}
A.from = () => {}
A.to = () => {}
// Output: (empty)

// Class declaration + property assignment → eliminated
class A {}
A.foo = 1
// Output: (empty)

// Const variable with arrow function → eliminated
const A = () => {}
A.foo = 1
// Output: (empty)

// Const variable with object literal → eliminated
const B = {}
B.foo = 1
// Output: (empty)

// Const variable with function expression → eliminated
const C = function() {}
C.foo = 1
// Output: (empty)
```

### How it works

1. Wires the existing `property_write_side_effects` from `oxc_ecmascript`'s `MayHaveSideEffectsContext` through the minifier's compress options
2. For member expression assignments, checks `may_have_side_effects()` then `is_member_assign_to_unused_binding()`
3. `is_member_assign_to_unused_binding` verifies:
   - Single-level member expression only (`A.foo`, not `a.b.c`)
   - All references to the symbol have the `MemberWriteTarget` flag (from parent PR)
   - Symbol creates a fresh value (`is_fresh_value`) and is not exported
4. Fixed-point loop handles cascading: first pass drops assignments, `exit_program` cleans up references, second pass drops declarations

### Fresh value detection

A symbol is considered "fresh" (cannot alias another binding and property writes are side-effect-free) if it is:
- A **function declaration**
- A **class declaration** without static setters, static accessor properties, or static property definitions with values
- A variable initialized with an **array/arrow function/function expression** literal
- A variable initialized with an **object literal** without setter/getter properties or property values containing setters/getters
- A variable initialized with a **class expression** (same rules as class declarations above)

Non-fresh values (e.g. `const b = a`) are conservatively preserved to prevent breaking aliased references.

#### Setter/getter detection (addresses review feedback)

Following how esbuild, Terser, Rollup, and SWC handle setter cases — all tools preserve property writes when setters are visible in the AST:

- **Static setter methods**: `class A { static set foo(v) { ... } }` → not fresh
- **Static accessor properties**: `class A { static accessor foo = 0 }` → not fresh (auto-generates getter+setter)
- **Static property definitions with values**: `class A { static b = 0 }` → not fresh (matches SWC's conservative approach)
- **Static property values containing setters/getters**: `class A { static b = { set x(v) { ... } } }` → not fresh
- **Object literal setters/getters**: `{ set foo(v) { ... } }` or `{ get foo() { ... } }` → not fresh
- **Nested setters/getters in object property values**: `{ bar: { set x(v) { ... } } }` → not fresh
- **Non-static setters are fine**: `class A { set foo(v) { ... } }` → fresh (instance setters don't affect class-level property writes)

### Limitations

**Only single-level member expressions are supported** (`A.foo`, not `a.b.c`). Deep nested assignments like `a.b.c = 1` are always preserved because intermediate properties could alias external objects:

```js
const exported = {};
const a = { b: exported };
a.b.c = 1; // writes to exported.c — unsafe to drop!
export { exported };
```

Supporting deeper nesting would require full object graph tracking (like Rollup's `ObjectEntity` model), which is a much larger effort for marginal gain.

### Safety guards

Addresses concerns raised in [PR #20730 review](https://github.com/oxc-project/oxc/pull/20730#pullrequestreview-4011376657) and [comment](https://github.com/oxc-project/oxc/pull/20730#issuecomment-4131710117):

- **Aliases**: `const b = a; b.add = 1; export { a }` → kept (not fresh, could mutate exported `a` through alias)
- **Chained access**: `const b = { a }; b.a.add = 1; export { a }` → kept (only single-level member expressions allowed)
- **Exported symbols**: `export function A() {} A.foo = 1;` → kept (observable by importers)
- **Other reads**: `function A() {} A.foo = 1; console.log(A)` → kept (symbol has non-member-write references)
- **Side-effectful RHS**: `function A() {} A.foo = sideEffect()` → kept (`may_have_side_effects` returns true)
- **Globals**: `globalObj.foo = 1` → kept (no local binding)
- **Script mode top-level**: top-level bindings in script mode are kept (they're globals visible to other scripts)
- **Setters**: `class A { static set foo(v) { ... } } A.foo = 1` → kept (setter triggers side effects)
- **Static properties**: `class A { static b = 0 } A.b = 1` → kept (matches SWC's conservative approach)

### Comparison with other tools

| Tool | Approach | Option needed? |
|------|----------|---------------|
| **Rollup** | Structural aliasing analysis — tracks object identity through `LocalVariable.isReassigned` + `ObjectEntity` setter detection. Drops `A.foo = 1` automatically when `A` is a local unreassigned object with no setters. | No (always active) |
| **Rolldown** | Adds explicit [`treeshake.propertyWriteSideEffects`](https://rolldown.rs/reference/InputOptions.treeshake#propertywritesideeffects) option (default `'always'`). Not in upstream Rollup. | Yes |
| **Terser** | Checks `is_constant_expression` on the base object after unwrapping property chain. If the base is constant within scope, drops the assignment. | No (always active, but less precise) |
| **esbuild** | All property assignments always preserved. Rationale: `Object.defineProperty` could install prototype setters. Only escape hatch is `/* @__PURE__ */` IIFE wrapping. | No option available |
| **SWC** | `may_have_side_effects` returns `true` for all assignments unconditionally. Classes with static property definitions are conservatively preserved. | No option available |
| **Oxc (this PR)** | Fresh value tracking (`is_fresh_value`) + `MemberWriteTarget` reference flags + explicit option. Closest to Rollup's aliasing analysis but gated behind an option like Rolldown. | Yes (`property_write_side_effects: false`) |

Closes #14207

## Test plan

- [x] 20 test cases covering all safety scenarios including setter/getter edge cases
- [x] All minifier tests pass (476/476)
- [x] No minsize regressions
- [x] Clippy clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Dunqing** [comment]
* **#20773** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20773?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20773?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#20772** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20772?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent changes, fast-track this PR to the front of the merge queue

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=20773).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by Dunqing** [comment]
Hi @sapphi-red, I reimplemented https://github.com/oxc-project/oxc/pull/20730 in this PR and solved two cases you mentioned in that. Please take a look again. Thank you!

**Comment by sapphi-red** [comment]
> * **Side-effectful RHS**: `function A() {} A.foo = sideEffect()` → kept (`may_have_side_effects` returns true)

We only need to keep the `sideEffect()` for this case.

> **SWC**	`may_have_side_effects` returns `true` for all assignments unconditionally. No optimization for property writes.

I think this is incorrect. SWC does remove them ([playground](https://play.swc.rs/?version=1.15.24&code=H4sIAAAAAAAAA0vOzysuUUjLz1ewVaiu5QIy9JISi4AcQy4Ave6ENhsAAAA%3D&config=H4sIAAAAAAAAA32VyW4bMQxA7%2F4KQ%2BceiqDooR8Q5NJvEGSJGsvVMhApx4PA%2F17O5jgxx7cRHxeJ23zs9nt1Qqv%2B7D%2F4kw%2B9qQj1dmYJDpnMhSUKbDJoa%2BhJ%2FVjpCUfkTUSYRNeZqFgKwkoWWQo5%2BOHety2pr4B4J2OpqV1LkAm%2F2i%2BslvcRUG338kMpEUx%2BQrRBHTJBB1VybEuMpkfQZ1MFL%2BNNTQ1YpBAjbARO97X0Is8uUCiZYz5SB8ZpWxwIKFSwFM4gmXEsNsvIzxPeM2EHh9Z1Uz2%2FWcPZxGZIiAmXqSR8W8HrsQQk7VuWUjjDjRzMcEnud8vgdQVqNT%2FanUrIGzX5B8AZiAYxmwSS30nDcz9tWfunliF7blkaBM79Lb0yQ8dJ1SF4IbNjZqBSkKpZwTULY2atdJ0Fb6QPgwMN3nOvCK7xPZA9SkFp6KF4AXB9jZe6agb6NoUbfByIJ%2FiVX0lygy0aydBxm%2BKQDiU%2BCZCAjsU9UeBSUNnGlbfEpd%2FmLTvg1gAnqjScwOMS4AGgoiOQ1Bs8HuxRd7EcPtfEonC97dtkcvc5719WLlwIKq%2BYN4j9w45TxFMwBVaALz9ffqndYqlScW1yufwCxq6Y9X7fKa27e723Cvh3NVxku%2Bt%2FCxPAVFAGAAA%3D)).

File: crates/oxc_minifier/tests/peephole/remove_unused_expression.rs [hunk 1]
```diff
@@ -476,3 +478,76 @@ fn remove_unused_class_expression() {
...
+    };
+
+    // Issue #14207: drop function declarations with property assignments
+    test_options("function A() {} A.from = () => {};", "", &options);
+
+    // Function declaration + multiple property assignments
+    test_options("function A() {} A.foo = 1; A.bar = 2;", "", &options);
+
+    // Class declaration + property assignment
+    test_options("class A {} A.foo = 1;", "", &options);
```
Comment (sapphi-red) [inline crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:501]: Are these handled? (I guess no?)
```js
class A {
  static set foo(v) { console.log("set foo") }
}
A.foo = 1;
```
```js
class A {
  static set foo(v) { console.log("set foo") }
}
class B extends A {}
B.foo = 1;
```
```js
const foo = {
  set foo(v) { console.log("set foo") }
};
foo.foo = 1;
```

File: crates/oxc_minifier/tests/peephole/remove_unused_expression.rs [hunk 2]
```diff
@@ -476,3 +478,76 @@ fn remove_unused_class_expression() {
...
+
+    // Alias where nothing is exported: inlining resolves alias, then everything drops
+    test_options("const a = {}; const b = a; b.foo = 1;", "", &options);
+
+    // Alias where target is exported: must preserve the property write
+    test_options(
+        "const a = {}; const b = a; b.add = 1; export { a };",
+        "const a = {}, b = a; b.add = 1; export { a };",
+        &options,
+    );
```
Comment (sapphi-red) [inline crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:536]: I guess this works but let's add a test case, just in case.
```suggestion
    test_options(
        "const a = {}; const b = a; b.add = 1; export { a };",
        "const a = {}, b = a; b.add = 1; export { a };",
        &options,
    );
    test_options(
        "const a = {}; const b = a; b.add = 1; export { b };",
        "const b = {}; b.add = 1; export { b };",
        &options,
    );
    test_options(
        "const a = {}; const b = a; a.add = 1; export { b };",
        "const a = {}, b = a; a.add = 1; export { b };",
        &options,
    );
```

File: crates/oxc_minifier/tests/peephole/remove_unused_expression.rs (hunk 1 above)
Comment (sapphi-red) [reply crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:501]: I guess these are handled indirectly:
```js
const foo = () => {};
Object.defineProperty(foo, "foo", {
  set: (v) => { console.log("set foo") }
});
foo.foo = 1;
```
```js
const foo = [];
Object.defineProperty(foo, "foo", {
  set: (v) => { console.log("set foo") }
});
foo.foo = 1;
```

Comment (Dunqing) [reply crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:501]: For the `class` case, no! I sent a commit that detects all properties to see whether they have a static set (explicit `set` or no); if so, preserve it. For the following case with `Object.defineProperty`, we already handled this correctly, as `foo` has a read reference in the first argument of `Object.defineProperty`

**Comment by Dunqing** [comment]
> > **SWC**	`may_have_side_effects` returns `true` for all assignments unconditionally. No optimization for property writes.
> 
> I think this is incorrect. SWC does remove them ([playground](https://play.swc.rs/?version=1.15.24&code=H4sIAAAAAAAAA0vOzysuUUjLz1ewVaiu5QIy9JISi4AcQy4Ave6ENhsAAAA%3D&config=H4sIAAAAAAAAA32VyW4bMQxA7%2F4KQ%2BceiqDooR8Q5NJvEGSJGsvVMhApx4PA%2F17O5jgxx7cRHxeJ23zs9nt1Qqv%2B7D%2F4kw%2B9qQj1dmYJDpnMhSUKbDJoa%2BhJ%2FVjpCUfkTUSYRNeZqFgKwkoWWQo5%2BOHety2pr4B4J2OpqV1LkAm%2F2i%2BslvcRUG338kMpEUx%2BQrRBHTJBB1VybEuMpkfQZ1MFL%2BNNTQ1YpBAjbARO97X0Is8uUCiZYz5SB8ZpWxwIKFSwFM4gmXEsNsvIzxPeM2EHh9Z1Uz2%2FWcPZxGZIiAmXqSR8W8HrsQQk7VuWUjjDjRzMcEnud8vgdQVqNT%2FanUrIGzX5B8AZiAYxmwSS30nDcz9tWfunliF7blkaBM79Lb0yQ8dJ1SF4IbNjZqBSkKpZwTULY2atdJ0Fb6QPgwMN3nOvCK7xPZA9SkFp6KF4AXB9jZe6agb6NoUbfByIJ%2FiVX0lygy0aydBxm%2BKQDiU%2BCZCAjsU9UeBSUNnGlbfEpd%2FmLTvg1gAnqjScwOMS4AGgoiOQ1Bs8HuxRd7EcPtfEonC97dtkcvc5719WLlwIKq%2BYN4j9w45TxFMwBVaALz9ffqndYqlScW1yufwCxq6Y9X7fKa27e723Cvh3NVxku%2Bt%2FCxPAVFAGAAA%3D)).


You are right, our implementation now matches the SWC implementation

File: crates/oxc_minifier/tests/peephole/remove_unused_expression.rs [hunk 3]
```diff
@@ -476,3 +478,144 @@ fn remove_unused_class_expression() {
...
+fn test_property_write_side_effects() {
+    let options = CompressOptions {
+        unused: CompressOptionsUnused::Remove,
+        treeshake: TreeShakeOptions {
+            property_write_side_effects: false,
+            property_read_side_effects: PropertyReadSideEffects::None,
+            ..TreeShakeOptions::default()
+        },
+        ..CompressOptions::smallest()
+    };
```
Comment (sapphi-red) [inline crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:492]: I came up with another edge case. Other than these, it looks good to me!

```js
const a = {}
a.__proto__ = { set a(v) { console.log('setter') } }
a.a = 1
```
```js
class A {}
A.__proto__ = { set a(v) { console.log('setter') } }
A.a = 1
```
Note: these are not changing `Object.prototype`

Comment (Dunqing) [reply crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:492]: Good edge case! Added tests in ac0c7311c0.

When the binding is completely unused (no export), dropping everything is safe — the `__proto__` assignment and `a.a = 1` are all removed together, so the setter never gets installed and `console.log` never fires.

When exported, both the `__proto__` write and property write are preserved correctly:
```js
// input
const a = {}; a.__proto__ = { set a(v) { console.log('setter'); } }; a.a = 1; export { a };
// output
const a = {}; a.__proto__ = { set a(v) { console.log('setter'); } }, a.a = 1; export { a };
```

Comment (sapphi-red) [reply crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:492]: Shouldn't we keep them regardless of whether it's exported? SWC keeps
```js
const a = {}
a.__proto__ = { set a(v) { console.log('setter') } }
a.a = 1
```
but removes
```js
const a = {}
a.__proto__ = { set a(v) { console.log('setter') } }
```
```js
const a = {}
a.a = 1
```

Comment (Dunqing) [reply crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:492]: You are right! Fixed in db3bdac3b3. Now matches SWC behavior:

```js
// kept — __proto__ installs setter, a.a triggers it
const a = {}; a.__proto__ = { set a(v) { console.log('setter') } }; a.a = 1;

// dropped — __proto__ alone, setter never triggered
const a = {}; a.__proto__ = { set a(v) { console.log('setter') } };

// dropped — no __proto__, plain property write
const a = {}; a.a = 1;
```

Implementation: track symbols with `__proto__` member writes in `MinifierState`. When a symbol has both `__proto__` writes and other property writes, preserve all of them.

File: crates/oxc_minifier/src/peephole/remove_unused_expression.rs [hunk 4]
```diff
@@ -636,6 +645,83 @@ impl<'a> PeepholeOptimizations {
...
+    ) -> bool {
+        // Only handle single-level member expressions (A.foo, not a.b.c).
+        // Chained access like `b.a.foo = 1` may write through aliased properties.
+        // Also detect `__proto__` writes which change the prototype chain and can
+        // install setters that make subsequent property writes side-effectful.
+        let (object, is_proto_write): (&Expression<'a>, bool) = match &assign_expr.left {
+            AssignmentTarget::StaticMemberExpression(e) => {
+                (&e.object, e.property.name == "__proto__")
+            }
+            AssignmentTarget::ComputedMemberExpression(e) => (&e.object, false),
```
Comment (sapphi-red) [inline crates/oxc_minifier/src/peephole/remove_unused_expression.rs]: I think we need to treat it as `is_proto_write = true` in the unknown case like `a[b]`.

File: crates/oxc_minifier/src/peephole/remove_unused_expression.rs [hunk 5]
```diff
@@ -636,6 +645,83 @@ impl<'a> PeepholeOptimizations {
...
+        if is_proto_write {
+            let ref_count = ctx.scoping().get_resolved_reference_ids(symbol_id).len();
+            if ref_count > 1 {
+                ctx.state.proto_write_symbols.insert(symbol_id);
+                return false;
+            }
+            // Fall through to normal checks — __proto__ alone can be dropped.
+        }
+        // If this symbol has `__proto__` writes, don't drop any property writes
+        // — the `__proto__` assignment may have installed setters.
```
Comment (sapphi-red) [inline crates/oxc_minifier/src/peephole/remove_unused_expression.rs]: Does this work when the `__proto__` assignment exists later in the code?
```js
const obj = {}
f()
obj.a = 1

function f() {
  obj.__proto__ = { set a(v) { console.log("hello") } }
}
```

File: crates/oxc_minifier/tests/peephole/remove_unused_expression.rs (hunk 3 above)
Comment (sapphi-red) [reply crates/oxc_minifier/tests/peephole/remove_unused_expression.rs:492]: I also noticed that we need to handle:
```js
const a = { __proto__: { set a(v) { console.log('setter') } } }
a.a = 1
```

File: crates/oxc_minifier/src/peephole/remove_unused_expression.rs (hunk 4 above)
Comment (Dunqing) [reply crates/oxc_minifier/src/peephole/remove_unused_expression.rs]: Good point

File: crates/oxc_minifier/src/peephole/remove_unused_expression.rs (hunk 5 above)
Comment (Dunqing) [reply crates/oxc_minifier/src/peephole/remove_unused_expression.rs]: Our implementation doesn't work well for this kind of case, see https://github.com/oxc-project/oxc/blob/51211223858ab7da8aded178006748a40c0c4c8d/crates/oxc_minifier/tests/peephole/remove_unused_expression.rs#L663-L670

The basic reason is our implementation doesn't pre-scan the AST to find all property writes, so when we enter the `obj.a = 1` expression, we actually don't know whether there is a `obj.__proto__` that comes later.

This is a limitation for us. How should we handle this? Is it fine to leave it for now?

Comment (sapphi-red) [reply crates/oxc_minifier/src/peephole/remove_unused_expression.rs]: It's a really edge case so I think we can leave this for now. Let's create an issue after merging this PR so we don't forget about it.

**Review by sapphi-red — APPROVED** [review]
LGTM

I think we should test this option out in many places after we merge this and is released in Rolldown / Vite.

File: crates/oxc_minifier/src/peephole/remove_unused_expression.rs (hunk 5 above)
Comment (Dunqing) [reply crates/oxc_minifier/src/peephole/remove_unused_expression.rs]: Created in https://github.com/oxc-project/oxc/issues/21248

**Comment by Dunqing** [comment]
### Merge activity

* **Apr 9, 11:12 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Apr 9, 11:12 AM UTC**: `Dunqing` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
* **Apr 9, 11:18 AM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
